### PR TITLE
Estimate width of the request based on watchers count in P&F

### PR DESCRIPTION
**Pull request by wojtek-t**
/kind feature
/priority important-soon
/sig api-machinery

```release-note
NONE
```

/assign @tkashem

### CODE REVIEW

**Comment by dims** [comment]
@wojtek-t needs rebase :(

**Comment by wojtek-t** [comment]
This is waiting for https://github.com/kubernetes/kubernetes/pull/103240 to be merged - @tkashem - FYI

**Comment by wojtek-t** [comment]
I realized I didn't add @MikeSpreitzer to this - PTAL

/retest

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/mutating_work_estimator.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/mutating_work_estimator.go:78]: nit: typo `processing event`

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/mutating_work_estimator.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/mutating_work_estimator.go:75]: maybe i am getting this wrong? can you clarify on this? if we substitute `seats` with `watchCount/watchesPerSeat` doesn't it always evaluate to 1?

**Review by tkashem — COMMENTED** [review]
had a quick question about `layers`

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/mutating_work_estimator.go
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/mutating_work_estimator.go:75]: This is true only seats was originally < maximumSeats (modulo roundings). If it was higher and we decreased to maximumSeats, this is no longer true.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/mutating_work_estimator.go
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/mutating_work_estimator.go:78]: done

**Comment by MikeSpreitzer** [comment]
I doubt we want the work estimator for this to assign a width greater than 1 to a write request.  Remember that a request's cost is rectangular: duration X seats.  All the seats are counted as occupied while the mutation is happening, as well as later when the watch notifications are being sent.  I am sorry I was busy with other matters when the relevant detail was merged into the KEP.

I would rather start with saying that a write request's width is set without regard to watch notifications, and then the extra latency is set to cover the estimated work of those notifications given the width already determined.  What is the actual structure of this work in terms of really running goroutines?

I note that in the LIST support we focused our attention on the costs of serializations and deserializations.  However, I see concern for other logic appearing in the WATCH support.  Are we being appropriately consistent?

**Comment by wojtek-t** [comment]
> I would rather start with saying that a write request's width is set without regard to watch notifications, and then the extra latency is set to cover the estimated work of those notifications given the width already determined. What is the actual structure of this work in terms of really running goroutines?

This is exactly the trick - while processing a write request is roughly single-threaded, processing of watches is embarassingly parallelizable - we generally have O(1) [3 IIRC to be more specific] goroutines per watch. So while write request is bounded in terms of resources it may use, it's not the case for watch processing.
As such - I don't think that we should be sticking to 1 seats for all write requests.

In the current logic, I'm trying to make as many requests as possible to stick to the width=1 (we probably can tune that even better and I'm open to suggestions). However, for the ones with hundreds of thousands of watchers registered, this isn't a realistic assumption.

> I note that in the LIST support we focused our attention on the costs of serializations and deserializations. However, I see concern for other logic appearing in the WATCH support. Are we being appropriately consistent?

That's a good question - this time is not as much about serializations/deserializations (as we optimize it across watches and try to do that ~once), but it's similarly proportional to the amount of data to process (which mostly boils down to amount of data to send in this case).

In my opinion we're consistent, as for LISTs we're effectively also talking about amount of data to process. And the difference live in the coefficient - serialization/deserialization has a different coefficient than the "data filtering/sending" which we're trying to reflect in the cost function (although certainly requires better tuning).

**Comment by MikeSpreitzer** [comment]
Multiple goroutines per WATCH?  That sounds really dangerous!  Has this actually caused problems?

**Comment by wojtek-t** [comment]
> Multiple goroutines per WATCH? That sounds really dangerous! Has this actually caused problems?

In the old versions of golang where goroutine starting was more expensive. Not since it improved there.

**Comment by wojtek-t** [comment]
> In the old versions of golang where goroutine starting was more expensive. Not since it improved there.

That said - it's somewhat orthogonal to this PR. We may be able to reduce the constant factor from 3x, but we will never reduce it to less than 1 gorotuine per watch, because http server starts gorotuine for each served request.
So we're only talking about a constant multiplier here.

**Comment by MikeSpreitzer** [comment]
> we will never reduce it to less than 1 gorotuine per watch, because http server starts gorotuine for each served request

Does the work have to be done on that goroutine?

> > Multiple goroutines per WATCH? That sounds really dangerous! Has this actually caused problems?
> In the old versions of golang where goroutine starting was more expensive. Not since it improved there.

If we are not actually seeing problems due to these WATCH notifications then maybe we do not have a problem there that needs to be solved?

**Comment by wojtek-t** [comment]
> Does the work have to be done on that goroutine?

I'm not sure what you're actually suggesting. Processing all watches from a single (constant number of) goroutines? That will not scale well enough...

> If we are not actually seeing problems due to these WATCH notifications then maybe we do not have a problem there that needs to be solved?

That's also why we don't prioritize it :)

**Comment by MikeSpreitzer** [comment]
I have some colleagues who have seen problems due to load from WATCH requests, but:
1. they were due to the load from the initialization phase rather than the cruise phase (the notifications triggered by writes during the WATCH), and
2. they were all due to poor choices in coding controllers, which were found (when investigating said load problems) and fixed.

**Comment by wojtek-t** [comment]
> they were due to the load from the initialization phase rather than the cruise phase (the notifications triggered by writes during the WATCH), and

That's a different (partially addressed) problem already.

If you take a large cluster (say 5k nodes), large service (say thousands of pods in a single service) and generate high churn on those, you will get the effect of what I'm saying - it will overload your apiserver immediately. Because we need to send a number of (previously large Endpoints), now moderately large EndpointSlice object to every node (kube-proxy) in the system.

**Comment by MikeSpreitzer** [comment]
So, @wojtek-t , it looks like you are saying that we _do_ have a problem that needs to be solved.  @lavalamp mentioned Endpoints in Slack, did not explicitly mention EndpointSlices.

**Comment by MikeSpreitzer** [comment]
> Processing all watches from a single (constant number of) goroutines? That will not scale well enough...

I am not sure I understand this.  We process the writes in a limited number of goroutines.  With the notification consideration we are adding to APF, why would no constant number of goroutines be adequate to deliver the notifications?

**Comment by wojtek-t** [comment]
> So, @wojtek-t , it looks like you are saying that we do have a problem that needs to be solved. @lavalamp mentioned Endpoints in Slack, did not explicitly mention EndpointSlices.

I've seen it problematic with EndpointSlices too (though EndpointSlices certainly pushed the limits).

But there are other examples of that:
- CiliumEndpoints (and soon to be CiliumEndpointSlices) are exactly the same pattern
- I've seen cases where users install their all custom daemons that are e.g. watching all pods from all nodes
- etc.

If we really want to provide protection, we can't ignore watches.

**Comment by MikeSpreitzer** [comment]
If I understand correctly, the concern here is limiting the CPU utilization and network data transmission rate of the apiserver.  Our management will be lumpy, but hopefully average out to a tolerable rate over time.  It matters whether we are talking about very short term averages or very long ones or something in between.  What is the time scale of the averaging that matters, for CPU and for network?

What CPU regulator punishes a process with excessive CPU utilization?  cgroups _enforces_ a cap, right?

**Comment by lavalamp** [comment]
> What CPU regulator punishes a process with excessive CPU utilization?

The problem is that any such limiter applies globally, punishing all requests. Network bandwidth throttling is particularly disruptive.

To summarize: the desire is to severely punish / throttle write requests which we know will trigger a large number of watch events. The watch events are sent in parallel to the extent possible, but the physical resources have limits (obviously). Both a width penalty and extra latency are therefore appropriate.

**Comment by MikeSpreitzer** [comment]
Sorry, my words were not clear enough regarding CPU.  When a CPU limit is imposed by cgroups, for example, the limited process is not able to use more than the imposed limit.  An overload, due to the work of sending watch notifications or anything else, does not cause the process to be killed; rather, all the requested work gets done, it just takes longer to get done than it would if more CPU could be used.  OTOH, if some other manager would allow the apiserver to use a forbidden amount of CPU and then kill it in response, that would make "excessive" CPU usage a problem.

**Comment by wojtek-t** [comment]
> Our management will be lumpy, but hopefully average out to a tolerable rate over time. It matters whether we are talking about very short term averages or very long ones or something in between. What is the time scale of the averaging that matters, for CPU and for network?

As I mentioned yesterday - I'm much more interested in guaranteeing overload protection then ensuring perfect utilization.
In that sense, I'm interested in overload protection over short horizons, but fairness about potentially slightly longer.


>  OTOH, if some other manager would allow the apiserver to use a forbidden amount of CPU and then kill it in response, that would make "excessive" CPU usage a problem.

I'm not sure I'm following where you're heading. Even ignoring any cgroups or any other mechanism, the amount of CPU will be limited by the machine on which kube-apiserver is running. Every setup is slightly different but I don't think that really matters.
All matters is that we can assume that kube-apiserver cannot exceed X number of resources (e.g. cpu).

If we ignore watches, those kind of operate "without admission" and consuming resources for requests that were admitted. While initially it may just slow them down, with accumulated additional load - we may simply completely starve them later.


BTW - continuing on the thought that I threw during yesterday sync about:
- reserving all seats to execute the request
- adjusting (for fairness reasons) to ensure that we better reflect the actual cost
would be to change this line:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go#L836
with something like:
```
r.queue.virtualStart -= (qs.estimatedServiceSeconds) * float64(r.Seats()) - S
```

**Comment by MikeSpreitzer** [comment]
Regarding time scale, I am looking for something quantitative.  Are we concerned about CPU usage averaged over a minute, a second, a few minutes, a few seconds, what?

**Comment by MikeSpreitzer** [comment]
@wojtek-t : I do not see how watch notifications can starve mutations.  Eventually all the notifications will be sent and the mutations will resume, even if the golang scheduler is that hostile.

@lavalamp's comment about timeouts may be what's important here.  We do not want mutations to be held up for something approaching a minute.

**Comment by MikeSpreitzer** [comment]
The point of the above questions is to inquire whether we can take a simpler approach to regulation here.  Although it does not directly reflect the schedule of the handling costs, the following idea may provide adquate regulation of the mutation rate.  My questions above were about whether that will be adequate, and the answers seem fairly affirmative.

Suppose that we give a request some additional latency and no additional width, so that the added work accounts for the estimated watch notifications.  By making requests take longer, this reduces the rate at which they will be served.

**Comment by wojtek-t** [comment]
> Regarding time scale, I am looking for something quantitative. Are we concerned about CPU usage averaged over a minute, a second, a few minutes, a few seconds, what?

Something at the level of 1 second, I would say.

> @wojtek-t : I do not see how watch notifications can starve mutations. Eventually all the notifications will be sent and the mutations will resume, even if the golang scheduler is that hostile.

If I have a tens (or even small hundreds) gorotuines processing mutations and hundreds of thousands processing watch events, we can easily starve them (not in a sense they would never finish, but to the point they will end up timing out even).

> Suppose that we give a request some additional latency and no additional width, so that the added work accounts for the estimated watch notifications. By making requests take longer, this reduces the rate at which they will be served.

Say my mutating event triggerred 5k watch events (5k watchers). With a single seat, we would have to allocate probably order of  seconds (or even more) time to process. But that's not what happens in reality, where they actually are being processed as fast as they can.
If we have say 600 seats available (as in default configuration), then accepting 600 API calls with each generating 5k events would easily overload kube-apiserver and lead to completely unpredictable things happening (some of which Daniel already descibed above).

**Comment by lavalamp** [comment]
> I do not see how watch notifications can starve mutations. Eventually all the notifications will be sent and the mutations will resume, even if the golang scheduler is that hostile.

The problem is not the golang scheduler, the problem is network bandwidth. E.g. imagine 1000 nodes with 10 services each covering 1000 pods and you do a rolling restart. Every pod change results in a change to an Endpoints object; there are 10 such objects, and each change generates an event for a watcher on each of the 1000 nodes. That is O(nodes * services * pods) total work, 10,000,000 total notifications if there is no batching. A big endpoints object can be, say, 500kib. That is 5TiB of network traffic. (In practice, there is some batching, and even if not, the first few notifications slow everything down so much that there is effectively more batching-- if the system continues to work at all.)

I am not sure if the problem is that the network egress prevents any mutating requests from coming into apiserver, or if it just prevents apiserver from saying anything back, but it doesn't matter. If you trigger this condition, apiserver--in fact, everything on the VM it is on-- doesn't respond to most requests for a very long time.

EndpointSlices help but don't eliminate the problem. And users can create the problem in other ways with other resources if they try.

**Comment by MikeSpreitzer** [comment]
OK, so it looks like just extending the duration will make the handling cost too bursty.

**Comment by MikeSpreitzer** [comment]
The problem here is inflating non-rectangular handling costs to a rectangle for our management purposes.  There are two distinguishable bad consequences of that.
1. The virtual world stuff is for determining the order of dispatch in the real world.  Inflating may have an unfair impact on that.
2. In the real world, server utilization is managed by seats.  Inflating seat occupancy potentially decreases server utilization.  I say _potentially_ because these seats are actually being time-shared onto much smaller numbers of the actually limited resources (e.g., CPUs, locks); over a large range of seat occupancy, changing the concurrency does not (I expect) have any significant effect on throughput.

In the virtual world the relevant effect is easy to cancel, because the virtual world scheduling does not really have seats: each queue runs just one request at a time, until its total work gets done.  And most of the logic does not deal directly with time but rather with readings from the progress meter R(t), which is an accumulator of work done by each queue (they are all assumed to do work at the same rate, this is the equal-rather-than-fair bug).  There is a guess at dispatch time of how much that work will be

```
	queue.virtualStart += qs.estimatedServiceSeconds * float64(request.Seats())
```

and an adjustment when the execution finishes to an accurate accounting

```
		r.queue.virtualStart -= (qs.estimatedServiceSeconds - S) * float64(r.Seats())
```

Interestingly, those lines are already buggy: neither takes the extra latency into account.  This means that the extra latency currently has no effect on the fairness mechanism.  I think that is bad.

Let us introduce two new fields in the `WorkEstimate` struct, `InitialSeats` and `FinalSeats`, to represent the width in the two distinct phases of handling, and make `Seats()` return the max of those two.

I think that dispatch time should take the extra latency into account.  Both in the estimate request completion times, used to find the expected next-to-complete request, and in the calculation of the expected start R of the following request.

```
	queue.virtualStart += qs.estimatedServiceSeconds*float64(request.workEstimate.InitialSeats) + request.workEstimate.AdditionalLatency.Seconds()*float64(request.workEstimate.FinalSeats)

```

Then, when the request completes, the adjustment should account for the difference between guessed and actual first phases.

```
		r.queue.virtualStart -= (qs.estimatedServiceSeconds - S) * float64(r.workEstimate.InitialSeats)
```

As for the real-world effect on server utilization, I am not sure how important it will be to do something about that.  I suspect it will be important.

We can limit the relative magnitude of the effect by limiting the fraction of the inflated rectangle that is not really occupied seats.  That fraction is, assuming FinalSeats is no smaller than InitialSeats, `(FinalSeats-InitialSeats) * estimatedServiceSeconds / FinalSeats * (estimatedServiceSeconds + additionalLatency)`.  We can enforce this limit by limiting how large we allow FinalSeats to be; the larger additionalLatency is, the larger we are allowed to set FinalSeats.

**Comment by MikeSpreitzer** [comment]
Also: imposing a limit on the unused fraction of the real-world scheduling rectangle makes the initial guess at service duration more consequential.  Perhaps it is time to start auto-tuning that.  Perhaps a simple running average (per FlowSchema, or per queue, I am not sure which is better) of service duration would suffice for that.

**Comment by lavalamp** [comment]
I think the best way to make sure we support the use case is to make an integration (or e2e if not possible) test that exercises the scenario. The test should be general and not depend on a particular fix implementation.

Then we should merge the stupidest thing that fixes that test without breaking other scenarios. This unlocks a large amount of value (e.g. we can start removing client side rate limits).

Then I think Mike's concerns come to the top. Optimizing stuff in the presence of good, general tests is fun and profitable :)

In that middle state, while this is (likely) excessively punitive, I think we'd want users to configure this so that such requests went to priority levels that were focused around the resource in question (e.g. endpoints).

**Comment by wojtek-t** [comment]
@MikeSpreitzer - thanks

What you suggest is basically an extension of what I proposed at the end of: https://github.com/kubernetes/kubernetes/pull/103539#issuecomment-923812909

Which I'm completely fine with. We should probably make this in a separate PR though.

> As for the real-world effect on server utilization, I am not sure how important it will be to do something about that. I suspect it will be important.

I think it's not that important. I strongly believe that the effect of that won't be significant and in those rare cases where it will visibly decrease utilization we actually prefer to reduce utilization than suffer from overload.

**Comment by wojtek-t** [comment]
> I think it's not that important. I strongly believe that the effect of that won't be significant and in those rare cases where it will visibly decrease utilization we actually prefer to reduce utilization than suffer from overload.

@MikeSpreitzer 

To clarify this part a bit - my point was mainly that - as long as we don't have borrowing between priority levels, underutilization within a single priority level will generally be negligible compared to underutilization across different priority levels.
So I guess we will get to the point when this might appear to be worth solving - but this will not be the case before we implement borrowing between priority levels.

**Comment by MikeSpreitzer** [comment]
@wojtek-t : how do you come to that conclusion about relative magnitude of under-utilization?

@lavalamp : what would an integration or e2e test the checks under-utilization look like?

**Comment by lavalamp** [comment]
>  what would an integration or e2e test that checks under-utilization look like?

The test I had in mind is to verify there's not OVER-utilization. And to do that I would set up the test scenario, e.g. X watchers of Y things updated as fast as possible, and verify that the number of watch events generated is below some threshold.

I am not worried about underutilization, I think that can be optimized later. If what we end up with is grossly underutilized, I believe we can change constant factors until it works in practice, and optimize for real later.

**Comment by wojtek-t** [comment]
/retest

@MikeSpreitzer @tkashem - PTAL

**Review by MikeSpreitzer — CHANGES_REQUESTED** [review]

**Comment by MikeSpreitzer** [comment]
The behavioral testing is inadequate in this regard, but we can fix that in a separate PR.

**Review by wojtek-t — COMMENTED** [review]
@MikeSpreitzer - thanks for the review

comments applied - PTAL

**Review by wojtek-t — COMMENTED** [review]
@MikeSpreitzer - thanks for the review.
I applied your comments and I'm unholding, but it now needs re-lgtm 

/hold cancel

**Review by MikeSpreitzer — COMMENTED** [review]
/LGTM

**Comment by MikeSpreitzer** [comment]
I am looking at https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-kubemark-500-gce/1448021022367289344 but am not familiar with how these tests work.

Also, BTW, why does the kube-apiserver log have entries like
```
I1012 21:07:32.046122      11 httplog.go:124] "HTTP" verb="GET" URI="/api/v1/namespaces/test-q781vv-1/secrets?allowWatchBookmarks=true&fieldSelector=metadata.name%3Daccount-53-11&resourceVersion=351459&timeout=7m13s&timeoutSeconds=433&watch=true" latency="1.721721704s" userAgent="kubemark/v1.23.0 (linux/amd64) kubernetes/af2ed25" audit-ID="efb7b8c4-d43b-477c-80e8-fd009ef4aabb" srcIP="10.64.5.26:35890" apf_pl="exempt" apf_fs="exempt" apf_fd="" apf_init_latency="296.835µs" resp=200
```
as well as entries like
```
I1012 21:07:31.983202      11 httplog.go:124] "HTTP" verb="WATCH" URI="/apis/storage.k8s.io/v1/csidrivers?allowWatchBookmarks=true&resourceVersion=327520&timeout=5m29s&timeoutSeconds=329&watch=true" latency="5m29.000606811s" userAgent="kubemark/v1.23.0 (linux/amd64) kubernetes/af2ed25" audit-ID="6aaa6416-4309-4097-a9c9-b6158845ba41" srcIP="10.64.7.46:55858" apf_pl="exempt" apf_fs="exempt" apf_fd="" apf_init_latency="303.958µs" resp=200
```
?

**Comment by MikeSpreitzer** [comment]
The complaint seems to be (with added whitespace for readability):
```
[measurement call APIResponsivenessPrometheus - APIResponsivenessPrometheusSimple error: top latency metric: there should be no high-latency requests, but: [
got: &{Resource:pods Subresource: Verb:POST Scope:resource Latency:perc50: 175.750111ms,
           perc90: 1.243873873s, perc99: 3.048313253s Count:18570 SlowCount:2401};
expected perc99 <= 1s

got: &{Resource:pods Subresource:binding Verb:POST Scope:resource Latency:perc50: 179.31098ms,
           perc90: 738.232044ms, perc99: 2.895361216s Count:18578 SlowCount:1343};
expected perc99 <= 1s

got: &{Resource:subjectaccessreviews Subresource: Verb:POST Scope:resource Latency:perc50: 35.03937ms,
          perc90: 144.285714ms, perc99: 1.554999999s Count:179 SlowCount:3};
expected perc99 <= 1s

got: &{Resource:events Subresource: Verb:PATCH Scope:resource Latency:perc50: 45.394736ms,
          perc90: 184.264705ms, perc99: 1.2575s Count:901 SlowCount:10};
expected perc99 <= 1s]]
```

I looked in the kube-apiserver log, and indeed there are some 3+ second latencies for POST to pods, almost entirely due to the APF filter.  Examples:

I1012 20:42:18.811476      11 httplog.go:124] "HTTP" verb="POST" URI="/api/v1/namespaces/test-dm2gcw-5/pods" latency="3.019339532s" userAgent="kube-controller-manager/v1.23.0 (linux/amd64) kubernetes/af2ed25/system:serviceaccount:kube-system:job-controller" audit-ID="26ece8e6-7844-493b-a7fd-9a2d8bf29908" srcIP="[::1]:46178" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_fd="test-dm2gcw-5" fl_priorityandfairness="3.017415416s" resp=201

I1012 20:42:18.818386      11 httplog.go:124] "HTTP" verb="POST" URI="/api/v1/namespaces/test-dm2gcw-5/pods" latency="3.01605656s" userAgent="kube-controller-manager/v1.23.0 (linux/amd64) kubernetes/af2ed25/system:serviceaccount:kube-system:job-controller" audit-ID="97a9b2a1-9bb3-461f-9003-55316faa470d" srcIP="[::1]:46178" apf_pl="workload-high" apf_fs="kube-system-service-accounts" apf_fd="test-dm2gcw-5" fl_priorityandfairness="3.014098725s" resp=201

**Comment by MikeSpreitzer** [comment]
Looking at https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-kubemark-500-gce/1448021022367289344/artifacts/MetricsForE2E_load_2021-10-12T21:02:06Z.json , I see

```
      {
        "metric": {
          "__name__": "apiserver_flowcontrol_priority_level_request_count_watermarks_bucket",
          "le": "0.5",
          "mark": "high",
          "phase": "executing",
          "priority_level": "workload-high"
        },
        "value": [
          0,
          "1722454"
        ]
      },
```

and all the higher buckets have the same value --- i.e., all the observations were less than or equal to 0.5.  In other words, the workload-high priority level never used more than half of its allocated concurrency according to this metric.  Now, to double-check what that metric really reports nowadays...

**Comment by wojtek-t** [comment]
Yeah - I think it's the culprit.
Let' me disable the main part with dropping the main part of the PR and we will work on renabling independently.

**Comment by MikeSpreitzer** [comment]
How about just tweaking up the constants?
I am finding lots of things to improve in the metrics.

**Pull request merged**
